tools/node-modules

```
@@ -0,0 +1,173 @@
+
+Options:
+
+    - tools/node-modules checkout --force     # disable this check
+
+    - tools/node-modules rebuild              # npm install with our package.json
+
+$0: *** aborting
+
+EOF
+            exit 1
+        fi
+    fi
+
+    # we're actually going to do this; let's remove the old one
+    cmd_remove
+
+    # and check out the new one
+    # we need to use the tag name here, unfortunately
+    [ "${V-}" != 0 ] || v0 UNPACK "node_modules  [sha: ${sha}]"
+    git -c advice.detachedHead=false clone ${quiet} "${CACHE_DIR}" -b "${tag}" node_modules
+}
+
+cmd_rebuild() {
+    for file in npm-shrinkwrap.json package-lock.json yarn.lock; do
+        if [ -e "${top_srcdir}/${file}" ]; then
+            echo "Please delete ${file} from ${top_srcdir} and start again"
+            exit 1
+        fi
+    done
+
+    cmd_remove
+
+    npm install
+
+    cp package.json node_modules/.package.json
+    echo /.cache >> node_modules/.gitignore
+
+    git -C node_modules init -b main
+    git -C node_modules add .
+    git -C node_modules -c gc.autoDetach=false commit -a -m "$(sha256sum package.json)"
+
+    cat <<EOF
+Next steps:
+
+  - tools/node-modules push
+
+  - git add node_modules && git commit
+
+EOF
+}
+
+cmd_push() {
+    local tag="sha-$(git -C node_modules rev-parse HEAD)"
+    git -C node_modules tag -f "${tag}" HEAD
+    git -C node_modules push "${PUSH_REMOTE}" "${tag}"
+}
+
+# called from Makefile.am
+cmd_make_package_lock_json() {
+    # Run from make to ensure package-lock.json is up to date
+
+    # package-lock.json is used as the stamp file for all things that use
+    # node_modules (mostly webpack), so this is the main bit of glue that
+    # drives the entire process
+
+    # We try our best not to touch package-lock.json unless it actually changes
+
+    # This isn't going to work for a tarball, but as long as
+    # package-lock.json is already there, we're OK
+    if [ ! -e .git -a -e package-lock.json ]; then
+        exit 0
```

This means that if someone is trying to work on an unpacked release tarball, this will silently succeed and ignore attempts to change package.json. ISTM that this should have a proper error message and exit 1 if package.json is newer.

---

Will add.